### ISSUE TITLE
IfcDoorStyle to IfcDoorType

### ISSUE THREAD

**Issue by ppaawweeuu** [IfcOpenShell/IfcOpenShell#4622]
Can I change IfcDoorStyle object (migrated from ifc2x3) to IfcDoorType (new in ifc4)?
Is it possible in BlenderBIM?

I use ifc schema, but file was upgraded from ifc2x3. I have in Types Collection few IfcDoorStyle objects and I cannot create IfcDoor instacne, and don't see any way to convert that IfcDoorStyle to IfcDoorType.

official [IfcDoorStyle](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/) spec states:
DEPRECATION  The entity is deprecated and shall not be used. The new entity [IfcDoorType](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/schema/ifcsharedbldgelements/lexical/ifcdoortype.htm) shall be used instead.

**Comment by ppaawweeuu** [comment]
I see that I can create IfcDoor using IfcDoorType, and change Relating Type Class to IfcDoorStyle. It is some kind workaround but it is probably ok, to avoid creating new doors using IfcDoorStyle. 

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/61344631/c82e6f0a-dda0-45b2-8438-71e53740bdff)

**Comment by theoryshaw** [comment]
Here's a proposed fix: https://github.com/IfcOpenShell/IfcOpenShell/pull/4623

Not 100% sure if there are any repercussions to this. Defer to others more knowledgeable.

**Comment by ppaawweeuu** [comment]
I see @theoryshaw proposal #4623, thanks a lot. I run BB wit that change. It solve my issue perfectly, but..

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/61344631/cec586bf-c051-4d3b-ad5c-0bb43e9706d0)

...I must admit that the more I think about that issue the more I like the way it works now.

That change may generate some new design process issues, e.g we can easyly change IfcDoorType to IfcDoorStyle, or just create new IfcDoorStyle from scratch.
If IFC4 standard states that we should avoid that class then we shouldn't allow user to easyly create new deprecated class.

btw: in my first post i edited link to [IfcDoorStyle](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/schema/ifcarchitecturedomain/lexical/ifcdoorstyle.htm) reference.

To summarise: I would leave that as it is. I can create new IfcDoor based on IfcDoorType or change relating class it to IfcDoorStyle.

**Comment by ppaawweeuu** [comment]
..on one hand deprecation is not prohibition, Decision in up to user.

..on other hand:
[here](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/link/annex-f.htm):
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/61344631/faa41c9e-d5a5-4652-b1dc-74d0e402f3ba)
and [here](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/schema/ifcarchitecturedomain/lexical/ifcdoorstyle.htm):
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/61344631/ed8bd27b-5eec-4261-a612-1d5f07b8672f)

_Defer to others more knowledgeable._

**Comment by theoryshaw** [comment]
I don't think the PR had any bearing on the following, that is, for ifc4 in BB, I think having the option of either IfcDoorStyle or IfcDoorType was the case before this PR.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/61344631/cec586bf-c051-4d3b-ad5c-0bb43e9706d0)

Nonetheless, it makes sense to not offer IfcDoorStyle in IFC4. 
As i understand it, to be strict about the following requirement, any IfcDoorStyles should be converted to IfcDoorTypes, upon a new save/export.  But maybe i'm missing something. 

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/507113/bd48473c-dbf8-43d1-b040-60061b9f067e)

**Comment by ppaawweeuu** [comment]
> I don't think the PR had any bearing

oops, you are right.

**Comment by Gorgious56** [comment]
In programming at least "Deprecated" means that something can still be used but is officially replaced by a new definition. It is still allowed to use this definition for a finite amount of subsequent versions, so as not to break workflows that depend on it overnight. But the goal is to eliminate the definition altogether in a more or less distant future, to alleviate technical debt. If you create a 4x3 project the option isn't available. 

So the solution would be to upgrade to Ifc4, convert to IfcDoorType, then upgrade to Ifc4x3 ?

**Comment by Andrej730** [comment]
@theoryshaw  #4623 is adding IfcTypeProduct to product type dropdown and it becomes kind of duplicate of IfcElementType option:

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/f8c9fef6-63d7-4ab4-9dbb-15573090e9d0)

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/fdfbd8f1-6199-4816-9a0b-05a4ed384222)

It seems that we can just change `can_reassign_class` to check something like `if element.is_a() in ("IfcWindowStyle", "IfcDoorStyle"): return True` to support editing classes for those elements, as they are kind of special.
https://github.com/IfcOpenShell/IfcOpenShell/blob/8f7f6223de6dbe6287652fed12b7a8a46db2ca4f/src/blenderbim/blenderbim/bim/module/root/data.py#L182-L190


Regarding deprecated types, I think it would make sense if something in the UI would indicate that class is deprecated and should be avoided but we shouldn't excluding it completely as it is available in IFC4.

> So the solution would be to upgrade to Ifc4, convert to IfcDoorType, then upgrade to Ifc4x3 ?

It's also possible that ifc4->4x3 migrator can handle conversion from IfcDoorStyles to IfcDoorTypes.

**Comment by theoryshaw** [comment]
Thanks @Andrej730 for the guidance!
